File: features/document/src/main/java/com/nivasafinance/features/document/service/impl/DocumentReadServiceImpl.java
package com.nivasafinance.features.document.service.impl;

import com.nivasafinance.features.document.dto.DocumentFileResponse;
import com.nivasafinance.features.document.dto.DocumentResponse;
import com.nivasafinance.features.document.entity.Document;
import com.nivasafinance.features.document.repository.DocumentRepositoryWrapper;
import com.nivasafinance.features.document.service.DocumentReadService;
import com.nivasafinance.features.document.storage.ContentRepository;
import com.nivasafinance.features.document.storage.ContentRepositoryFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.InputStream;
import java.util.UUID;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class DocumentReadServiceImpl implements DocumentReadService {
    
    private final DocumentRepositoryWrapper documentRepositoryWrapper;
    private final ContentRepositoryFactory contentRepositoryFactory;
    
    @Override
    public DocumentResponse getDocumentById(Long id) {
        return DocumentResponse.from(documentRepositoryWrapper.findByIdWithException(id));
    }
    
    @Override
    public DocumentResponse getDocumentByIdentifier(UUID id) {
        return DocumentResponse.from(documentRepositoryWrapper.findByIdentifierWithException(id));
    }
    
    @Override
    public DocumentFileResponse getDocumentFile(Long id) {
        Document document =
                documentRepositoryWrapper.findByIdWithException(id);
        return buildDocumentFileResponse(document);
    }

    @Override
    public DocumentFileResponse getDocumentFile(UUID id) {
        Document document =
                documentRepositoryWrapper.findByIdentifierWithException(id);
        return buildDocumentFileResponse(document);
    }

    private DocumentFileResponse buildDocumentFileResponse(Document document) {
        ContentRepository contentRepository =
                contentRepositoryFactory.getRepository(document.getProvider().name());
        InputStream file = contentRepository.fetchFile(document.getPath());
        return DocumentFileResponse.builder()
                .file(file)
                .data(DocumentResponse.from(document))
                .build();
    }
}

File: features/document/src/main/java/com/nivasafinance/features/document/service/impl/DocumentReadServiceImpl.kt
package com.nivasafinance.features.document.service.impl

import com.nivasafinance.features.document.dto.DocumentFileResponse
import com.nivasafinance.features.document.dto.DocumentResponse
import com.nivasafinance.features.document.dto.toDocumentResponse
import com.nivasafinance.features.document.repository.DocumentRepositoryWrapper
import com.nivasafinance.features.document.service.DocumentReadService
import com.nivasafinance.features.document.storage.ContentRepositoryFactory
import com.nivasafinance.features.master.codemaster.SystemControlledMasterCodes.DOCUMENT_MASTER
import com.nivasafinance.features.master.codemaster.service.CodeMasterService
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.UUID

@Service
@Transactional(readOnly = true)
class DocumentReadServiceImpl(
    private val documentRepositoryWrapper: DocumentRepositoryWrapper,
    private val contentRepositoryFactory: ContentRepositoryFactory
) : DocumentReadService {
    override fun getDocumentById(id: Long): DocumentResponse {
        return documentRepositoryWrapper.findByIdWithException(id).toDocumentResponse()
    }

    override fun getDocumentByIdentifier(id: UUID): DocumentResponse {
        return documentRepositoryWrapper.findByIdentifierWithException(id).toDocumentResponse()
    }

    override fun getDocumentFile(id: UUID): DocumentFileResponse {
        val document = documentRepositoryWrapper.findByIdentifierWithException(id)
        val contentRepository = contentRepositoryFactory.getRepository(document.provider.name)
        val file = contentRepository.fetchFile(document.path)
        return DocumentFileResponse(file = file, data = document.toDocumentResponse())
    }
}
